### PR TITLE
Reworking leveldb code

### PR DESCRIPTION
**Pull request by 0xcb9ff9**
# Description

* fix missing state storage leveldb configuration
* state storage performance issues
* unified kv storage abstract interface

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [x] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Breaking changes

leveldb builder api move to `helper/kvdb/`, `leveldb.DB` implement  `KVBatchStorage` interface

# Checklist

- [ ] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [ ] I have added the relevant labels
- [ ] I have updated the official documentation
- [ ] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [ ] I have tested this code manually

### CODE REVIEW

**Comment by DarianShawn** [comment]
It is `refactor`, and `backward-compatible`, so it is not a `breaking change`.

File: helper/kvdb/kvdb.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package kvdb
+
+type KVBatch interface {
+	Set(k, v []byte)
+	Write()
```
Comment (DarianShawn) [inline helper/kvdb/kvdb.go]: Must returns error when failed. We should not "swallow" any failure.

File: blockchain/storage/leveldb/builder.go [hunk 2]
```diff
@@ -1,156 +1,30 @@
...
-			OpenFilesCacheCapacity: minHandles,
-			CompactionTableSize:    DefaultCompactionTableSize * leveldbopt.MiB,
-			CompactionTotalSize:    DefaultCompactionTotalSize * leveldbopt.MiB,
-			BlockCacheCapacity:     minCache / 2 * leveldbopt.MiB,
-			WriteBuffer:            minCache / 4 * leveldbopt.MiB,
-			Filter:                 filter.NewBloomFilter(DefaultBloomKeyBits),
-			NoSync:                 false,
-		},
+// NewBlockchainStorageBuilder creates the new blockchain storage builder
+func NewBlockchainStorageBuilder(logger hclog.Logger, leveldbBuilder kvdb.LevelDBBuilder) blockchain.StorageBuilder {
```
Comment (DarianShawn) [inline blockchain/storage/leveldb/builder.go]: Should it be a high level API instead of the inner module?

**Review by DarianShawn — COMMENTED** [review]
Great job, man. 👍🏻
Keep them consistent is a valuable thing.
A few more suggestions to make it better.

File: state/immutable-trie/trie.go [hunk 3]
```diff
@@ -194,7 +194,11 @@ func (t *Trie) Commit(objs []*state.Object) (state.Snapshot, []byte) {
 	nTrie.storage = t.storage
 
 	// Write all the entries to db
-	batch.Write()
+	// TODO, need to handle error
```
Comment (0xcb9ff9) [inline state/immutable-trie/trie.go:198]: @DarianShawn add this error handling to the to-do list, the scope of the change API is too large

File: blockchain/storage/leveldb/builder.go (hunk 2 above)
Comment (0xcb9ff9) [reply blockchain/storage/leveldb/builder.go]: @DarianShawn  only one place to use, maybe move to up

Comment (0xcb9ff9) [reply blockchain/storage/leveldb/builder.go]: @DarianShawn can't move to the upper level, exist cycle import

File: helper/kvdb/kvdb.go (hunk 1 above)
Comment (0xcb9ff9) [reply helper/kvdb/kvdb.go]: @DarianShawn 9d3272f

File: state/immutable-trie/trie.go (hunk 3 above)
Comment (DarianShawn) [reply state/immutable-trie/trie.go:198]: Got it.
We should separate it to some small PRs.

File: helper/kvdb/kvdb.go (hunk 1 above)
Comment (DarianShawn) [reply helper/kvdb/kvdb.go]: Got it. 🙌

**Comment by DarianShawn** [comment]
I've updated the base branch. Should merge back to `dev` branch after the PR fully tested in some RPC nodes.

**Comment by 0xcb9ff9** [comment]
@DarianShawn ok, rebase to `reuion-leveldb-interface`, commit 54cf9d2 refactoring blocakchain storage interface

File: state/immutable-trie/storage.go [hunk 4]
```diff
@@ -17,81 +16,39 @@ var (
...
-	Put(k, v []byte)
-	Write()
-}
-
 // Storage stores the trie
 type Storage interface {
-	Put(k, v []byte)
-	Get(k []byte) ([]byte, bool)
-	Batch() Batch
+	kvdb.KVBatchStorage
```
Comment (DarianShawn) [inline state/immutable-trie/storage.go]: How about keeping those functions? It was a 'duck type' contraction. 

As a matter of fact, high-level modules use abstractions directly from low-level modules, this approach violates the principle of dependency inversion (DIP).

I would suggest that we keep those interface, which implementation could be replaced at any moment.

File: state/immutable-trie/storage.go [hunk 5]
```diff
@@ -17,81 +16,39 @@ var (
...
-// KVBatch is a batch write for leveldb
-type KVBatch struct {
-	db    *leveldb.DB
-	batch *leveldb.Batch
-}
-
-func (b *KVBatch) Put(k, v []byte) {
-	b.batch.Put(k, v)
+type kvStorage struct {
+	kvdb.KVBatchStorage
```
Comment (DarianShawn) [inline state/immutable-trie/storage.go]: Don't ever nest interface to struct, it was a bad practice. Especially use low module interface.

You could verify interface compliance using [uber-go golang guide](https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance), which compiler could do those verification job for you.

File: command/server/server.go [hunk 6]
```diff
@@ -67,42 +67,42 @@ func setFlags(cmd *cobra.Command) {
 	cmd.Flags().IntVar(
 		&params.leveldbCacheSize,
```
Comment (DarianShawn) [inline command/server/server.go]: Consider rename those flags to `--db.cache, --db.xxx` or `--kvdb.xxx` ?
These word feel more like the same level as the 'network', etc.

File: blockchain/storage/kvstorage/kvstorage.go [hunk 7]
```diff
@@ -48,23 +49,14 @@ var (
 	EMPTY  = []byte("empty")
 )
 
-// KV is a key value storage interface.
-//
-// KV = Key-Value
```
Comment (DarianShawn) [inline blockchain/storage/kvstorage/kvstorage.go:53]: Keep those interface please. Use lower module interface or API violate DIP (Dependency Inversion Principle).

**Review by DarianShawn — COMMENTED** [review]
Appreciate with your great job. 🙌 

To make it better, I left some review comments. Give any feed back if you need help.

**Review by DarianShawn — APPROVED** [review]
Great job. It is definitely what it should be. 
Thanks for your contribution.

**Comment by 0xcb9ff9** [comment]
@DarianShawn rebase to current branch 1606b36

File: command/server/server.go (hunk 6 above)
Comment (DarianShawn) [reply command/server/server.go]: It was a recommendation, not a request change.
So I'll set the conversation to resolved.

**Pull request merged**